odl/util/utility.py

```diff
@@ -147,3 +128,89 @@ def __new__(cls, name, this_bases, d):
                 return type.__new__(cls, name, (), d)
             return meta(name, bases, d)
     return metaclass('temporary_class', None, {})
+
+
+def is_scalar_dtype(dtype):
+    """Whether a datatype is scalar or not."""
+    return np.issubsctype(dtype, np.number)
+
+
+def is_int_dtype(dtype):
+    """Whether a datatype is integer or not."""
+    return np.issubscdtype(dtype, np.integer)
+
+
+def is_real_dtype(dtype):
+    """Whether a datatype is real (including integer) or not."""
+    return is_scalar_dtype(dtype) and not is_complex_floating_dtype(dtype)
+
+
+def is_real_floating_dtype(dtype):
+    """Whether a NumPy datatype is real complex-floating or not."""
+    return np.issubsctype(dtype, np.floating)
+
+
+def is_complex_floating_dtype(dtype):
+    """Whether a NumPy datatype is complex floating-point or not."""
+    return np.issubsctype(dtype, np.complexfloating)
+
+
+def is_valid_input_array(x, d):
+    """Test whether `x` is a correctly shaped array of points in R^d."""
+    if not isinstance(x, np.ndarray):
+        return False
+    if d == 1:
+        return x.ndim == 1 or x.ndim == 2 and x.shape[0] == 1
+    else:
+        return x.ndim == 2 and x.shape[0] == d
+
+
+def is_valid_input_meshgrid(x, d):
+    """Test whether `x` is a meshgrid sequence for points in R^d."""
+    if isinstance(x, np.ndarray):
+        return False
+    if d > 1:
+        try:
+            np.broadcast(*x)
+        except ValueError:  # cannot be broadcast
+            return False
+
+    return (len(x) == d and
+            all(isinstance(xi, np.ndarray) for xi in x) and
+            all(xi.ndim == d for xi in x))
+
+
+def meshgrid_input_order(x):
+    """Determine the ordering of a meshgrid argument."""
+    # Case 1: all elements have the same shape -> non-sparse
+    if all(xi.shape == x[0].shape for xi in x):
+        # Contiguity check only works for meshgrid created with copy=True.
+        # Otherwise, there is no way to find out the intended ordering.
+        if all(xi.flags.f_contiguous for xi in x):
+            return 'F'
+        else:
+            return 'C'
```

Comment:
Internally this can never happen, it occurs only if a user manually creates a meshgrid with `numpy.meshgrid(*x, sparse=False, copy=False)`. In that case, it is hard to do anything since you cannot find out the user's initial intention and need to assume something or raise an exception. Assuming `'C'` ordering is okay IMO, since it's by far the most common case. Needs to be documented in that case, though. Raising would be the safe variant.